File: server_python/models/ticket.py
from enum import Enum
from datetime import datetime
from typing import Optional, List, Dict, Any
from uuid import UUID, uuid4
from pydantic import BaseModel, Field


class TicketStatus(str, Enum):
    DRAFT = "draft"
    PENDING_APPROVAL = "pending_approval"
    APPROVED = "approved"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    REJECTED = "rejected"
    BLOCKED = "blocked"


class Priority(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"


class TicketOption(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    label: str
    description: str
    isRecommended: bool = False
    metadata: Optional[Dict[str, Any]] = None


class TicketSourceType(str, Enum):
    EMAIL = "email"
    DOCUMENT = "document"
    CHAT = "chat"
    CALENDAR = "calendar"
    MANUAL = "manual"
    SYSTEM = "system"


class TicketSource(BaseModel):
    type: TicketSourceType
    reference: Optional[str] = None


class Ticket(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    agentId: str
    
    # 작업 정의 (5W1H 기반)
    purpose: str  # Why - 왜 이 작업이 필요한가
    content: str  # What - 무엇을 할 것인가
    context: Optional[str] = None  # Where/When - 맥락 정보
    
    # 사용자 결정 요청
    decisionRequired: Optional[str] = None
    options: List[TicketOption] = Field(default_factory=list)
    selectedOptionId: Optional[str] = None
    
    # 실행 계획
    executionPlan: str
    estimatedImpact: Optional[str] = None
    
    # 메타데이터
    status: TicketStatus = TicketStatus.DRAFT
    priority: Priority = Priority.MEDIUM
    
    # 연관 정보
    parentTicketId: Optional[str] = None
    childTicketIds: List[str] = Field(default_factory=list)
    relatedTicketIds: List[str] = Field(default_factory=list)
    
    # 추적 정보
    source: Optional[TicketSource] = None
    
    # 제약조건 (온톨로지에서 적용된)
    appliedConstraints: List[str] = Field(default_factory=list)
    
    # 타임스탬프
    createdAt: datetime = Field(default_factory=datetime.now)
    updatedAt: datetime = Field(default_factory=datetime.now)
    approvedAt: Optional[datetime] = None
    completedAt: Optional[datetime] = None


class CreateTicketInput(BaseModel):
    agentId: str
    purpose: str
    content: str
    context: Optional[str] = None
    decisionRequired: Optional[str] = None
    options: Optional[List[TicketOption]] = None
    executionPlan: str
    estimatedImpact: Optional[str] = None
    priority: Priority = Priority.MEDIUM
    parentTicketId: Optional[str] = None
    childTicketIds: Optional[List[str]] = None
    relatedTicketIds: Optional[List[str]] = None
    source: Optional[TicketSource] = None
    appliedConstraints: Optional[List[str]] = None

File: server/src/models/ticket.ts
import { z } from 'zod';

// 티켓 상태
export const TicketStatusSchema = z.enum([
  'draft',              // 초안 - Agent가 작성 중
  'pending_approval',   // 승인 대기 - 사용자 결정 필요
  'approved',           // 승인됨 - 실행 가능
  'in_progress',        // 진행 중 - Agent가 실행 중
  'completed',          // 완료
  'rejected',           // 거부됨
  'blocked',            // 차단됨 - 추가 정보 필요
]);

export type TicketStatus = z.infer<typeof TicketStatusSchema>;

// 우선순위
export const PrioritySchema = z.enum(['low', 'medium', 'high', 'urgent']);
export type Priority = z.infer<typeof PrioritySchema>;

// 티켓 옵션 (사용자 선택지)
export const TicketOptionSchema = z.object({
  id: z.string().uuid(),
  label: z.string().min(1),
  description: z.string(),
  isRecommended: z.boolean().default(false),
  metadata: z.record(z.unknown()).optional(),
});

export type TicketOption = z.infer<typeof TicketOptionSchema>;

// 티켓 스키마 (작업 단위)
export const TicketSchema = z.object({
  id: z.string().uuid(),
  agentId: z.string(),

  // 작업 정의 (5W1H 기반)
  purpose: z.string().min(1),           // Why - 왜 이 작업이 필요한가
  content: z.string().min(1),           // What - 무엇을 할 것인가
  context: z.string().optional(),       // Where/When - 맥락 정보

  // 사용자 결정 요청
  decisionRequired: z.string().optional(),  // 필요한 결정 사항
  options: z.array(TicketOptionSchema).default([]),
  selectedOptionId: z.string().uuid().optional(),

  // 실행 계획
  executionPlan: z.string(),            // 승인 후 수행할 내용
  estimatedImpact: z.string().optional(), // 예상 영향도

  // 메타데이터
  status: TicketStatusSchema.default('draft'),
  priority: PrioritySchema.default('medium'),

  // 연관 정보
  parentTicketId: z.string().uuid().optional(),  // 상위 티켓 (분해된 경우)
  childTicketIds: z.array(z.string().uuid()).default([]),
  relatedTicketIds: z.array(z.string().uuid()).default([]),

  // 추적 정보
  source: z.object({
    type: z.enum(['email', 'document', 'chat', 'calendar', 'manual', 'system']),
    reference: z.string().optional(),
  }).optional(),

  // 제약조건 (온톨로지에서 적용된)
  appliedConstraints: z.array(z.string()).default([]),

  // 타임스탬프
  createdAt: z.date(),
  updatedAt: z.date(),
  approvedAt: z.date().optional(),
  completedAt: z.date().optional(),
});

export type Ticket = z.infer<typeof TicketSchema>;

// 티켓 생성 입력
export const CreateTicketInputSchema = TicketSchema.omit({
  id: true,
  createdAt: true,
  updatedAt: true,
  approvedAt: true,
  completedAt: true,
  status: true,
}).partial({
  childTicketIds: true,
  relatedTicketIds: true,
  appliedConstraints: true,
  options: true,
});

export type CreateTicketInput = z.infer<typeof CreateTicketInputSchema>;
